industrial_taxonomy/pipeline/glass_house/flow.py

```diff
@@ -0,0 +1,196 @@
+
+    @step
+    def start(self):
+        """Load raw data"""
+        from industrial_taxonomy.getters.companies_house import (
+            get_name as get_companies_house_names,
+        )
+        from industrial_taxonomy.getters.glass import (
+            get_organisation as get_glass_organisation,
+        )
+
+        self.run_id = current.origin_run_id or current.run_id
+
+        nrows = 10_000 if (self.test_mode and not current.is_production) else None
+
+        self.names_x = get_companies_house_names().name.head(nrows)
+        self.names_y = get_glass_organisation().name.head(nrows)
+
+        self.next(self.clean_names)
+
+    @pip(path="requirements.txt")
+    @step
+    def clean_names(self):
+        """Pre-process names if `self.should_clean_names`."""
+        from jacc_hammer.name_clean import preproc_names
+
+        self.clean_names_x = (
+            preproc_names(self.names_x) if self.should_clean_names else self.names_x
+        )
+        self.clean_names_y = (
+            preproc_names(self.names_y) if self.should_clean_names else self.names_y
+        )
+
+        self.next(self.match)
+
+    @resources(memory=64_000)
+    @pip(path="requirements.txt")
+    @step
+    def match(self):
+        """The core fuzzy matching algorithm"""
+        from itertools import starmap as starmap_
+        from pathlib import Path
+        from tempfile import TemporaryDirectory
+
+        from jacc_hammer.fuzzy_hash import (
+            Cos_config,
+            Fuzzy_config,
+            match_names_stream,
+        )
+        from pandas import DataFrame
+        from toolz.curried import curry, pipe
+        from toolz.curried import map as tmap
+
+        from utils import optimise_chunk
+
+        starmap = curry(starmap_)
+
+        def persist_chunk(chunk_id: int, chunk: DataFrame) -> int:
+            key = f"match-chunks-{chunk_id}"
+            with S3(run=self) as s3:
+                s3.put(key, chunk.to_csv())
+            print(f"Persisted chunk {chunk_id} of shape {chunk.shape}")
+            return chunk_id
+
+        tmp_dir = Path(TemporaryDirectory(dir=".").name)
+        tmp_dir.mkdir()
+
+        cos_config = Cos_config()
+        fuzzy_config = Fuzzy_config(num_perm=128)
+        match_config = dict(
+            threshold=50,
+            chunksize=100_000,
+            cos_config=cos_config,
+            fuzzy_config=fuzzy_config,
+            tmp_dir=tmp_dir,
+        )
+
+        self._chunk_ids: List[int] = pipe(
+            match_names_stream(
+                [self.clean_names_x.tolist(), self.clean_names_y.tolist()],
+                **match_config,
+            ),
+            tmap(optimise_chunk),
+            enumerate,  # Give chunks ID's
+            starmap(persist_chunk),
+            list,  # Realise computation
+        )
+        self.next(self.extract_top_matches)
+
+    @resources(memory=64_000)
+    @step
+    def extract_top_matches(self):
```

Comment:
Is there any contingency or notification for if a chunk fails or isn't present for some reason/unexpected AWS behaviour?

Reply:
Then the flow will fail. Do you have any ideas?

Adding `@retry` can solve transient errors but because this flow isn't run regularly and when it does run it takes ~12 hours then I decided failing earlier and kicking up a fuss was better than having to wait another ~12 hours in the case that it wasn't a transient error.

Reply:
I think the flow failing is a good solution. I was worried about e.g. the flow running and one machine prematurely terminating and losing some of the results. I guess it would make sense that Metaflow takes this into account!

Reply:
To clarify: This step is only running on one machine

The matching algorithm has to be performed on the same machine

We can't chunk the input data but it is possible to chunk the output data.

Reply:
Rather than pickling one big dataframe and saving it in the datastore

We save lots of smaller chunks as separate s3 objects because we can produce a chunk, save it, and then free that memory to be used by the next chunk

Reply:
Got it! Thanks